---
title: "Brett Report"
author: "Karl Heilbron"
output: html_document
params:
  maindir: ""
  ld.panel: ""
  gw.file: ""
  chr.bp.col: ""
  chr.col: ""
  bp.col: ""
  a1.col: ""
  a2.col: ""
  p.col: ""
  eaf.col: ""
  n1.col: ""
  n0.col: ""
  n.col: ""
  n: ""
  z.or.p: ""
  check.args: ""
---

```{r echo=FALSE}
# Set up shorter parameter names
gw.file <- params$gw.file
```

# Inputs
```{r echo=FALSE}

# Print inputs
library(DT)
input_df <- data.frame( Argument=names(params), Description="" )
for( i in seq_len( NROW(input_df) ) ){
  if( !is.null(params[[i]]) ){
    input_df$Description[i] <- params[[i]]
  }
}
datatable( data    = input_df,
           caption = "Inputs used to generate this report" )

# Show the first 5 rows of the GWAS file
library(data.table)
gw <- fread( file=gw.file, nrows=5 )
datatable( data    = gw, 
           caption = "The first 5 rows of the GWAS file" )
```

# GWAS peaks
```{r echo=FALSE}
peaks <- fread( file=file.path( maindir, "" ) )
names(peaks) <- toupper( names(peaks) )
peak_caption <- paste( "A table of GWAS hits. SNP = variant identifier,",
                        "CHR = chromosome, BP = position, LO = leftmost",
                        "boundary of the peak, HI = rightmost boundary of",
                        "the peak, P = P value" )
datatable( data    = peaks, 
           caption = peak_caption )
```


