---
title: "Brett Report"
author: "Karl Heilbron"
output: 
  html_document:
    toc: true
    toc_float: true
params:
  maindir: ""
  ld.panel: ""
  gw.file: ""
  chr.bp.col: ""
  chr.col: ""
  bp.col: ""
  a1.col: ""
  a2.col: ""
  p.col: ""
  eaf.col: ""
  n1.col: ""
  n0.col: ""
  n.col: ""
  n: ""
  z.or.p: ""
  check.args: ""
---

```{r simplify_variable_names, echo=FALSE}
# Set up shorter parameter names
gw.file <- params$gw.file
```

# Inputs {.tabset .tabset-pills}
## Inputs
```{r inputs, echo=FALSE}

# Print inputs
library(DT)
input_df <- data.frame( Argument=names(params), Description="" )
for( i in seq_len( NROW(input_df) ) ){
  if( !is.null(params[[i]]) ){
    input_df$Description[i] <- params[[i]]
  }
}
datatable( data    = input_df,
           caption = "Inputs used to generate this report" )
```

## GWAS file
```{r gwas_file, echo=FALSE}

# Show the first 5 rows of the GWAS file
library(data.table)
gw <- fread( file=gw.file, nrows=5 )
datatable( data    = gw, 
           caption = "The first 5 rows of the GWAS file" )
```

# GWAS peaks
```{r gwas_peaks, echo=FALSE}
peaks <- fread( file=file.path( maindir, "peaks.tsv" ) )
names(peaks) <- toupper( names(peaks) )
peak_caption <- paste( "A table of GWAS hits. SNP = variant identifier,",
                        "CHR = chromosome, BP = position, LO = leftmost",
                        "boundary of the peak, HI = rightmost boundary of",
                        "the peak, P = P value" )
datatable( data    = peaks, 
           caption = peak_caption )
```

# Top genes
```{r top_genes, echo=FALSE, warning=FALSE}

# Read in POPS results, calculate P values
pops_file <- file.path( maindir, "pops.preds" )
pops <- fread(pops_file)
names(pops)[ names(pops) == "PoPS_Score" ] <- "POPS"
pops$P_POPS <- z_to_p( z=pops$POPS )

# Read in MAGMA results
mag_file <- file.path( maindir, "magma.genes.out" )
mag      <- fread(mag_file)

# Read in gene locations
gene_file <- "/home/heilbron/projects/pops/data/gene_locations.tsv"
genes     <- fread(gene_file)

# Add gene names and MAGMA results to POPS results
pops$GENE  <- genes$NAME[  match( pops$ENSGID, genes$ENSGID ) ]
pops$CHR   <- genes$CHR[   match( pops$ENSGID, genes$ENSGID ) ]
pops$START <- genes$START[ match( pops$ENSGID, genes$ENSGID ) ]
pops$STOP  <- genes$END[   match( pops$ENSGID, genes$ENSGID ) ]
pops$P_MAGMA <- mag$P[     match( pops$ENSGID, mag$GENE ) ]
pops$Z_MAGMA <- mag$ZSTAT[ match( pops$ENSGID, mag$GENE ) ]
  
# Subset to interesting columns
gcols <- c( "GENE", "CHR", "START", "STOP", 
            "POPS", "Z_MAGMA", "P_POPS", "P_MAGMA" )
genes_caption <- paste( "A table of top genes based on POPS and MAGMA results.",
                        "GENE = HGNC gene name, CHR = chromosome, START =",
                        "leftmost gene boundary, STOP = rightmost gene boundary,",
                        "POPS = POP score, Z_MAGMA = MAGMA z-score, P_POPS =",
                        "P value associated with the POP score, P_MAGMA =",
                        "MAGMA P value" )
datatable( data     = pops[ , ..gcols ],
           caption  = genes_caption,
           rownames = FALSE,
           options  = list( order = list( list( 4, "desc" ) ) ) ) %>%
  formatRound( columns = c( 'POPS', 'Z_MAGMA', 'P_POPS' ), 
               digits  = 3 )
```

# POPS and MAGMA distributions
## POPS
```{r pops_distribution, echo=FALSE}
plot( density(pops$POPS) )
```

## MAGMA
```{r magma_distribution, echo=FALSE}
plot( density(pops$Z_MAGMA) )
```





