---
title: "Brett Report"
author: "Karl Heilbron"
output: 
  html_document:
    toc: true
    toc_float: true
params:
  maindir: ""
  ld_panel: ""
  gw_file: ""
  loci_dir: ""
  check_args: ""
---

```{r simplify_variable_names, echo=FALSE}
# Set up shorter parameter names
gw_file  <- params$gw_file
maindir  <- params$maindir
loci_dir <- params$loci_dir
```

# Inputs {.tabset .tabset-pills}
## Inputs
```{r inputs, echo=FALSE}

# Print inputs
library(DT)
input_df <- data.frame( Argument=names(params), Description="" )
for( i in seq_len( NROW(input_df) ) ){
  if( !is.null(params[[i]]) ){
    input_df$Description[i] <- params[[i]]
  }
}
datatable( data    = input_df,
           caption = "Inputs used to generate this report" )
```

## GWAS file
```{r gwas_file, echo=FALSE}

# Show the first 5 rows of the GWAS file
library(data.table)
gw <- fread( file=gw_file, nrows=5 )
datatable( data    = gw, 
           caption = "The first 5 rows of the GWAS file" )
```

# POPS and MAGMA distributions {.tabset .tabset-pills}
## POPS
```{r pops_distribution, echo=FALSE}

# Read in POPS results, calculate P values
pops_file <- file.path( maindir, "pops.preds" )
pops <- fread(pops_file)
names(pops)[ names(pops) == "PoPS_Score" ] <- "POPS"
pops$P_POPS <- z_to_p( z=pops$POPS )
pops$Q_POPS <- ( rank(pops$POPS) - 1 ) / ( NROW(pops) -  1)

# Read in MAGMA results
mag_file <- file.path( maindir, "magma.genes.out" )
mag      <- fread(mag_file)

# Read in gene locations
gene_file <- "/projects/0/prjs0817/projects/pops/data/gene_locations.tsv"
genes     <- fread(gene_file)

# Add gene names and MAGMA results to POPS results
pops$GENE  <- genes$NAME[  match( pops$ENSGID, genes$ENSGID ) ]
pops$CHR   <- genes$CHR[   match( pops$ENSGID, genes$ENSGID ) ]
pops$START <- genes$START[ match( pops$ENSGID, genes$ENSGID ) ]
pops$STOP  <- genes$END[   match( pops$ENSGID, genes$ENSGID ) ]
pops$P_MAGMA <- mag$P[     match( pops$ENSGID, mag$GENE ) ]
pops$Z_MAGMA <- mag$ZSTAT[ match( pops$ENSGID, mag$GENE ) ]

# Density plot  
plot( density(pops$POPS), main="", las=1, lwd=2, col="grey30" )
```

## MAGMA
```{r magma_distribution, echo=FALSE}

# Density plot
plot( density( na.omit(pops$Z_MAGMA) ), main="", las=1, lwd=2, col="grey30" )
```

## POPS v. MAGMA
```{r pops_v_magma, echo=FALSE}
plot( x=pops$Z_MAGMA, y=pops$POPS, main="", las=1, col="grey30",
      xlab="MAGMA z-score", ylab="POP Score" )
```

# GWAS peaks
```{r gwas_peaks, echo=FALSE}

# Read in peaks
pk_file <- file.path( maindir, "peaks.tsv" )
peaks <- fread(pk_file)
names(peaks) <- toupper( names(peaks) )

# Print table
gcols_peaks <- c( "SNP", "CHR", "CENTRE", "LO", "HI", "P", "D_GENE", "DIST", 
                  "P_GENE", "POPS", "N_GENES", "BOTH", "PRIORITY" )
peak_caption <- paste( "A table of GWAS hits.",
                       "SNP = variant identifier,",
                       "CHR = chromosome,",
                       "CENTRE = PIP-weighted position,",
                       "LO = leftmost boundary of the peak,",
                       "HI = rightmost boundary of the peak,",
                       "P = top-PIP SNP P value,",
                       "D_GENE = gene nearest to the PIP-weighted position,",
                       "DIST = distance between PIP-weighted position and the",
                       "nearest gene, ",
                       "P_GENE = gene with the highest POPS in the locus,",
                       "POPS = POPS score,",
                       "N_GENES = number of genes in the locus,",
                       "BOTH = the nearest gene is also the top POPS gene,",
                       "PRIORITY = the nearest gene is also the top POPS gene",
                       "and the POPS score is in the top 5% of all scores and",
                       "there are <= 12 genes in the locus." )
datatable( data    = peaks[ , ..gcols_peaks ], 
           caption = peak_caption,
           filter  = list( position = 'top', clear = FALSE ) ) %>% 
  formatRound( columns = c( 'POPS' ), 
               digits  = 3 ) %>%
  formatSignif( columns = c('P'), digits = 2 )
```

# Top genes
```{r top_genes, echo=FALSE, warning=FALSE}

# Subset to interesting columns
gcols <- c( "GENE", "CHR", "START", "STOP", 
            "POPS", "Z_MAGMA", "P_POPS", "P_MAGMA" )

# Print table
genes_caption <- paste( "A table of top genes based on POPS and MAGMA results.",
                        "GENE = HGNC gene name, CHR = chromosome, START =",
                        "leftmost gene boundary, STOP = rightmost gene boundary,",
                        "POPS = POP score, Z_MAGMA = MAGMA z-score, P_POPS =",
                        "P value associated with the POP score, P_MAGMA =",
                        "MAGMA P value" )
datatable( data     = pops[ , ..gcols ],
           caption  = genes_caption,
           rownames = FALSE,
           filter = list( position = 'top', clear = FALSE ),
           options  = list( order = list( list( 4, "desc" ) ) ) ) %>%
  formatRound( columns = c( 'POPS', 'Z_MAGMA', 'P_POPS' ), 
               digits  = 3 ) %>%
  formatSignif( columns = c('P_MAGMA'), digits = 2 )
```

# Locus plots
```{r locus_plots, echo=FALSE, results="asis"}

# Loop through peaks
# peaks <- peaks[ order( peaks$CHR, peaks$BP ) , ]
for( i in seq_along(peaks$HIT) ){
  
  # Set up the sub-heading for the locus
  locus_heading <- paste0( "\n## Locus ", i, ": ", peaks$D_GENE[i], " {.tabset .tabset-pills}\n")
  locus_desc <- paste0( peaks$SNP[i], 
                        ", Chr", peaks$CHR[i], " ", round(peaks$CENTRE[i]/1e6), 
                        "Mb, Near: ", peaks$D_GENE[i], 
                        ", PoPS: ", peaks$P_GENE[i], "\n" )
  cat(locus_heading)
  cat(locus_desc)
  cat("\n### CS-based loci")
  
  # Figure out how many hits are in the locus
  n_hits <- sum( peaks$LOCUS == peaks$LOCUS[i] )
  
  # If there are multiple hits, insert the unconditioned LZP
  if( n_hits > 1 ){
    lzp_uncond_plot <- file.path( loci_dir, "lzp", "separated", "unconditioned", 
                                  paste0( peaks$HIT[i], ".jpg" ) )
    cat( '\n<font size="5"> Marginal P values</font>' )
    cat( "\n", paste0( "![](", lzp_uncond_plot, ")" ), "\n" )
  }
  
  # Insert the conditioned LZP plot
  lzp_cond_plot <- file.path( loci_dir, "lzp", "separated", "conditioned", 
                                paste0( peaks$HIT[i], ".jpg" ) )
  cat( '\n<font size="5"> Conditional P values</font>' )
  cat( "\n", paste0( "![](", lzp_cond_plot, ")" ), "\n" )
  
  # Insert the POPS plot
  pops_locus_plot <- file.path( "plots", "pops", 
                                paste0( "region_", i, "_", peaks$SNP[i], ".jpg" ) )
  cat( "\n", paste0( "![](", pops_locus_plot, ")" ), "\n" )
  
  # Insert the MAGMA plot
  # mag_locus_plot <- file.path( "plots", "magma", 
  #                              paste0( "region_", i, "_", peaks$SNP[i], ".jpg" ) )
  # cat( "\n", paste0( "![](", mag_locus_plot, ")" ), "\n" )
  
  # If there are multiple hits, insert the unconditioned and conditioned merged LZPs
  if( n_hits > 1 ){
    cat("\n### LD-based loci")
    merged_uncond_plot <- file.path( loci_dir, "lzp", "merged", "unconditioned", 
                                  paste0( peaks$HIT[i], ".jpg" ) )
    merged_cond_plot   <- file.path( loci_dir, "lzp", "merged", "conditioned", 
                                  paste0( peaks$HIT[i], ".jpg" ) )
    cat( '\n <font size="5"> Marginal P values</font>' )
    cat( "\n", paste0( "![](", merged_uncond_plot, ")" ), "\n" )
    cat( '\n <font size="5"> Conditioned P values</font>' )
    cat( "\n", paste0( "![](", merged_cond_plot, ")" ), "\n" )
  }
}
```




